src/vsr/replica.zig

```diff
@@ -154,6 +154,14 @@ pub fn Replica(
         commit_max: u64,
 
         /// Whether we are reading a prepare from storage in order to commit.
+        /// Guards against concurrent commits.
+        ///
+        /// While committing, a replica may not:
+        ///
+        /// * send a `do_view_change` message, or
+        /// * start a view as the new leader
+        ///
+        /// because `commit_min` will be modified when the commit finishes.
         committing: bool = false,
```

Comment:
Unfortunately not. Right now the commit path works like:

1. Either:
    - Commit from WAL:
      1. `commit_ops`: Check `committing` guard. Either proceed or return.
      1. `commit_ops`: Set `committing = true`.
      1. `commit_ops_read`: Call `journal.read_prepare`.
      1. → Call `commit_op_hooks`.
    - Commit from pipeline:
      1. `commit_pipeline`: Check `committing` guard. Either proceed or return.
      1. `commit_pipeline`: Set `committing = true`.
      1. `commit_pipeline_next`: Call `commit_op_hooks`.
2. `commit_op_hooks`: Set `commit_callback` and `commit_prepare`, start prefetch
3. → prefetch done
4. start compact
5. → compact done, set `commit_callback = null` and `commit_prepare = null`, call (former) `commit_callback`.
6. The callback implementations (there are 2: one for commit-from-WAL, and one for commit-from-pipeline) will either:
    - decide to stop committing: call `commit_ops_done` (which sets `committing = false`)
    - or, jump to step 3 to commit another op.

(where "→" indicates that the step follows asynchronously).

If you find this hard to follow — I do too :sweat_smile:. But I'm leaning towards holding off refactoring for a separate PR to avoid holding up the LSM work.

One thing I think could help in the mean time is renaming some of the functions (suggestion: https://github.com/coilhq/tigerbeetle/pull/122#issuecomment-1199970696).

Reply:
100% agreed on a separate PR. Thanks for the detailed explanation here.